### ISSUE TITLE
Is it possible to train both a JAX and PyTorch model at the same time?

### ISSUE THREAD

**Issue by Artur-Galstyan** [google/torchax#26]
Hi everyone. I'm trying to use torchax to train a JAX model, that contains some PyTorch modules. I managed to train the JAX part, but the PyTorch part remains frozen. I'm using Equinox as my NN library. This is the  MVE:

```python
from typing import Any

import equinox as eqx
import jax
import jax.numpy as jnp
import optax
import torch
import torchax
from torchax.interop import JittableModule
from torchvision import datasets, transforms

torchax.enable_globally()

transform = transforms.Compose(
    [transforms.ToTensor(), transforms.Normalize((0.1307,), (0.3081,))]
)
train_dataset = datasets.MNIST(
    root="./data", train=True, download=True, transform=transform
)
train_loader = torch.utils.data.DataLoader(
    train_dataset, batch_size=32, shuffle=True, drop_last=True
)


class TorchPart(torch.nn.Module):
    def __init__(self, in_features, out_features):
        super().__init__()
        self.linear = torch.nn.Linear(in_features, out_features)

    def forward(self, x):
        return self.linear(x)


class HybridModel(eqx.Module):
    jax_linear1: eqx.nn.Linear
    torch_part: JittableModule
    jax_linear2: eqx.nn.Linear

    def __init__(self, key):
        self.jax_linear1 = eqx.nn.Linear(784, 128, key=key)
        torch_model = TorchPart(128, 128).to("jax")
        self.torch_part = JittableModule(torch_model)
        self.jax_linear2 = eqx.nn.Linear(128, 10, key=key)

    def __call__(self, x: jnp.ndarray) -> jnp.ndarray:
        x = x.reshape(-1, 784)
        x = eqx.filter_vmap(self.jax_linear1)(x)
        x = x.reshape(-1, 128)
        x = self.torch_part(x).jax()
        x = jax.nn.relu(x)
        x = eqx.filter_vmap(self.jax_linear2)(x)
        return x


def loss_fn(model: HybridModel, x: jnp.ndarray, y: jnp.ndarray) -> jnp.ndarray:
    logits = model(x)
    return optax.softmax_cross_entropy_with_integer_labels(
        logits=logits, labels=y
    ).mean()


def torchax_filter(element: Any):
    if eqx.is_array(element):
        return True
    if isinstance(element, torch.Tensor):
        return True
    return False


@eqx.filter_jit
def train_step(
    model: HybridModel,
    opt_state: Any,
    optimizer: optax.GradientTransformation,
    x: jnp.ndarray,
    y: jnp.ndarray,
) -> tuple[HybridModel, Any, jnp.ndarray]:
    (loss_val, grads) = eqx.filter_value_and_grad(loss_fn)(model, x, y)

    updates, opt_state = optimizer.update(grads, opt_state, model)
    model = eqx.apply_updates(model, updates)
    return model, opt_state, loss_val


main_key = jax.random.key(42)
model = HybridModel(main_key)

optimizer = optax.adam(1e-3)
opt_state = optimizer.init(eqx.filter(model, torchax_filter))

linear_before_training = model.torch_part.params["linear.weight"].jax()
print("before training", linear_before_training)


for epoch in range(1):
    for batch_idx, (x_torch, y_torch) in enumerate(train_loader):
        x_jax = jnp.array(x_torch.numpy())
        y_jax = jnp.array(y_torch.numpy())

        model, opt_state, loss = train_step(model, opt_state, optimizer, x_jax, y_jax)

        if batch_idx % 100 == 0:
            print(f"Batch {batch_idx}, Loss: {loss.item()}")

        if batch_idx >= 500:
            break


linear_after_training = model.torch_part.params["linear.weight"].jax()
print("after training", linear_after_training)
```

A snippet from the output shows e.g. this 

```
before training [[-0.05707549 -0.08610713 -0.04081073 ...  0.07019104 -0.05778148

...

after training [[-0.05707549 -0.08610713 -0.04081073 ...  0.07019104 -0.05778148
```

Which are the same numbers. But the loss is going down, so the JAX part is actually training. 

Is it even theoretically possible to train the PyTorch weights too? My initial guess as to why it's skipping the PyTorch weights is because `eqx.filter_value_and_grad` filters using the following:

```python
diff_x, nondiff_x = partition(x, is_inexact_array)
```

and `is_inexact_array` filters like this

```python
def is_inexact_array(element: Any) -> bool:
    """Returns `True` if `element` is an inexact (i.e. floating or complex) JAX/NumPy
    array.
    """
    if isinstance(element, (np.ndarray, np.generic)):
        return bool(np.issubdtype(element.dtype, np.inexact))
    elif isinstance(element, jax.Array):
        return jnp.issubdtype(element.dtype, jnp.inexact)
    else:
        return False
```

My feeling is that it should be possible somehow, but I just can't this to work :(

**Comment by qihqi** [comment]
Yes, your suspicion is correct. However, after working around it, likely it still won't work. So for `equinox` to work, it must know that a child module it has can become pytree flattenable into Jax arrays. 

In other words, JittableModule contains the weights, but equinox's Module don't know it also have weights.

My recommendation to keep the state that torch model has as plain pytree with Jax arrays, then when you need to call the torch module, do it functionally (passing in the weights as input).

Here is the modified script:

```python
from typing import Any

import equinox as eqx
import jax
import jax.numpy as jnp
import optax
import torch
import torchax
from torchax.interop import JittableModule
from torchvision import datasets, transforms

from torchax.interop import jax_view, call_torch

torchax.enable_globally()

transform = transforms.Compose(
    [transforms.ToTensor(), transforms.Normalize((0.1307,), (0.3081,))]
)
train_dataset = datasets.MNIST(
    root="./data", train=True, download=True, transform=transform
)
train_loader = torch.utils.data.DataLoader(
    train_dataset, batch_size=32, shuffle=True, drop_last=True
)


class TorchPart(torch.nn.Module):
    def __init__(self, in_features, out_features):
        super().__init__()
        self.linear = torch.nn.Linear(in_features, out_features)

    def forward(self, x):
        return self.linear(x)


class HybridModel(eqx.Module):
    jax_linear1: eqx.nn.Linear
    # torch_part: JittableModule
    jax_linear2: eqx.nn.Linear
    torch_weights: tuple
    _func_call_torch: Any


    def __init__(self, key):
        self.jax_linear1 = eqx.nn.Linear(784, 128, key=key)
        torch_model = TorchPart(128, 128).to("jax")
        torch_part = JittableModule(torch_model)
        # NOTE: jax_view makes a jax-pytree from a torch-pytree
        self.torch_weights = jax_view(torch_part.params)
        self.jax_linear2 = eqx.nn.Linear(128, 10, key=key)

        def _call_torch_functional(weights, x):
          # NOTE: buffer part is usually the non-trainable part of the model, so
          # here we pass in as closure and it wont change by training as equinox doesnt know it
          # If you want to train it then you can assign it to a attr like torch_weights, and
          # take it as input in this function
          return torch_part.functional_call('forward', weights, torch_part.buffers, x)
        self._func_call_torch = _call_torch_functional

    def __call__(self, x: jnp.ndarray) -> jnp.ndarray:
        x = x.reshape(-1, 784)
        x = eqx.filter_vmap(self.jax_linear1)(x)
        x = x.reshape(-1, 128)
        x = call_torch(self._func_call_torch, self.torch_weights, x)
        x = jax.nn.relu(x)
        x = eqx.filter_vmap(self.jax_linear2)(x)
        return x


def loss_fn(model: HybridModel, x: jnp.ndarray, y: jnp.ndarray) -> jnp.ndarray:
    logits = model(x)
    return optax.softmax_cross_entropy_with_integer_labels(
        logits=logits, labels=y
    ).mean()


def torchax_filter(element: Any):
    if eqx.is_array(element):
        return True
    if isinstance(element, torch.Tensor):
        return True
    return False


@eqx.filter_jit
def train_step(
    model: HybridModel,
    opt_state: Any,
    optimizer: optax.GradientTransformation,
    x: jnp.ndarray,
    y: jnp.ndarray,
) -> tuple[HybridModel, Any, jnp.ndarray]:
    (loss_val, grads) = eqx.filter_value_and_grad(loss_fn)(model, x, y)

    updates, opt_state = optimizer.update(grads, opt_state, model)
    model = eqx.apply_updates(model, updates)
    return model, opt_state, loss_val


main_key = jax.random.key(42)
model = HybridModel(main_key)

optimizer = optax.adam(1e-3)
opt_state = optimizer.init(eqx.filter(model, torchax_filter))

linear_before_training = model.torch_weights["linear.weight"]
print("before training", linear_before_training)


for epoch in range(1):
    for batch_idx, (x_torch, y_torch) in enumerate(train_loader):
        x_jax = jnp.array(x_torch.numpy())
        y_jax = jnp.array(y_torch.numpy())

        model, opt_state, loss = train_step(model, opt_state, optimizer, x_jax, y_jax)

        if batch_idx % 100 == 0:
            print(f"Batch {batch_idx}, Loss: {loss.item()}")

        if batch_idx >= 500:
            break


linear_after_training = model.torch_weights["linear.weight"]
print("after training", linear_after_training)
```

The output I got is 

```
before training [[ 0.08054759  0.04976586  0.08233765 ...  0.06200922 -0.03341131
   0.01342426]
 [ 0.0055548   0.01805575  0.03758043 ...  0.06117871 -0.03179939
  -0.05745878]
 [ 0.00133851 -0.00630595 -0.00205481 ...  0.01129127  0.04164737
  -0.07105379]
 ...
 [-0.0271943   0.06178992 -0.03555684 ... -0.00271146 -0.03920677
   0.05235033]
 [ 0.03833425 -0.06290725  0.0760347  ...  0.04804095  0.03361678
  -0.05714056]
 [ 0.02353179  0.05340045  0.03952353 ...  0.00325849 -0.05032492
   0.06526745]]
/Users/hanq/homebrew/Caskroom/miniconda/base/envs/py13/lib/python3.13/site-packages/jax/_src/numpy/lax_numpy.py:5946: UserWarning: Explicitly requested dtype int64 requested in arange is not available, and will be truncated to dtype int32. To enable more dtypes, set the jax_enable_x64 configuration option or the JAX_ENABLE_X64 shell environment variable. See https://github.com/jax-ml/jax#current-gotchas for more.
  return _arange(start, stop=stop, step=step, dtype=dtype,
Batch 0, Loss: 2.305910110473633
Batch 100, Loss: 0.3939344584941864
Batch 200, Loss: 0.12738938629627228
Batch 300, Loss: 0.16933700442314148
Batch 400, Loss: 0.12957459688186646
Batch 500, Loss: 0.6118622422218323
after training [[ 0.05713508  0.02663921  0.0466007  ...  0.11207806 -0.06381968
   0.0049739 ]
 [-0.01692384 -0.01071602  0.066591   ...  0.03805482 -0.02916645
  -0.0752075 ]
 [-0.0137838   0.00091874  0.00150477 ...  0.00339229  0.10176165
  -0.07887951]
 ...
 [-0.03891971  0.0504028  -0.02816711 ... -0.0065406  -0.08681227
   0.04183269]
 [ 0.03239078 -0.09252942  0.07566665 ...  0.02350532  0.0298689
  -0.04094533]
 [ 0.07871515  0.0777012   0.03873569 ... -0.0347164  -0.08190042
   0.11365014]]
```

**Comment by Artur-Galstyan** [comment]
This is great! Thank you for the help! I will tinker with this and investigate some more and if everything works, we can close this issue and I'll perhaps create a separate PR for an Equinox example, if wanted :)
